File: src/4-problem11.pas
// Nama/NIM  : Muhammad Alif Arifin/13516078
// Tema      : Array - Bab 4 Problem 11
// Deskripsi : Menghitung jumlah kemunculan di array
// Batasan   : - 
// Asumsi    : 1) NIM 004 dengan 04 sama 
//             2) jumlah input tidak lebih dari 100

program bab4;

{ KAMUS }
var
    nim : array[1..100] of smallint; // menyimpan nim
    n, i, j, count, temp : integer; // n = mencatat jumlah nim
                                    // i, j = untuk iterasi loop
                                    // count = untuk pencatatan
                                    // temp = untuk penampung sementara

begin
    write('Masukkan jumlah kertas kehadiran: ');
    readln(n);

    if (n > 100) then
        writeln('Maaf, jumlah maksimum adalah 100')
    // input valid (kalaupun non positif tidak akan masuk loop)
    else 
    begin
        // meminta input
        for i := 1 to n do
        begin
            // melakukan validasi
            repeat
                write('Masukkan NIM mahasiswa: ');
                readln(temp);
            until ((temp > 0) and (temp < 1000));
            nim[i] := temp;
        end;

        // melakukan pencatatan
        writeln('Rekap kehadiran mahasiswa:');
        for i := 1 to n do
        begin
            // jika nim tidak valid
            if (nim[i] <> -1) then
            begin                        
                count := 1;
                for j := (i + 1) to n do
                begin
                    if (nim[i] = nim[j]) then
                    begin
                        count := count + 1;
                        nim[j] := -1; // agar tidak diakses kembali
                    end;
                end;
                writeln(nim[i], ' - ', count, ' kali');
            end;
        end;
    end;
end.

File: src/4-Problem11.py
## Bab 4 Soal 11

n = int(input("Masukkan jumlah kertas kehadiran : ")) # asumsi masukan valid
nimCount = []

if (n != 0):
	x = int(input("Masukkan NIM mahasiswa : ")) # asumsi input sudah sesuai yang diinginkan
	nimCount.append(x)
	
for i in range(1, n):
	x = int(input("Masukkan NIM mahasiswa : "))
	idx = 0
	while (idx < i-1 and nimCount[idx] < x):
		idx += 1
	if (nimCount[idx] >= x):
		nimCount.insert(idx, x)
	else: # jika idx di elemen terakhir dan x > nimCount[idx]
		nimCount.append(x)

nimCount.append('a') # penanda akhir list
print "Rekap kehadiran mahasiswa :"
x = 0 # penanda counter
	
i = 1
while (i < n):
	while (nimCount[i] == nimCount[x]): # inkremen i jika nim sama
		i += 1
	print "%s - %d kali" % (nimCount[x], i - x) # i - x adalah berapa banyak kemunculan nim sama
	x = i

# Penjelasan:
# Maaf tak mengerti fungsi-fungsi python yang lebih baru.
# Pada awalnya ingin menggunakan dictionary, tetapi peserta belum tahu cara
# mensortir nim yang ditampilkan, sehingga beralih ke list.
# Peserta mensortir masukan pada saat dimasukkan, dengan mencari list sampai ditemukan
# posisi yang tepat. Maaf walau judulnya array, peserta tak dapat menemukan array
# di tutorial python. 
# Suatu dummy element ('a') ditambahkan ke akhir list untuk menghentikan proses pengecekan
# apakah ada nim sama atau berbeda. Awal masukan diambil masukan pertama dahulu, agar
# program tak perlu mengecek list kosong atau diisi. Program akan membaca list mulai
# dari suatu nim, lalu membaca berapa kali nim sama muncul dalam list sampai ditemukan
# nim berbeda atau dummy element. Asumsi print: satu newline saja.
# Peserta berpikir menggunakan array 2-dimensi (pertama untuk nim, kedua untuk
# jumlah kehadiran) tidak cocok karena bab 5 bertopik matriks. Maaf penjelasan
# soal ini lebih banyak maafnya.
# (Peserta juga tidak paham mengapa kertas yang ada 10, tetapi program hanya
# meminta 9 kertas)
